src/lib/Types/Core.hs

```diff
@@ -392,25 +401,32 @@ data Cache (n::S) = Cache
 
 -- === object file representations ===
 
--- TODO: think about the naming discipline for compiled function names. We can't
--- use our standard naming systems because these need to interact with the
--- outside world, LLVM etc.
-
-type CFunName = String
-data CFun n = CFun CFunName (ObjectFileName n)
-              deriving (Show, Generic)
-
-type ObjectFileName = Name ObjectFileNameC
-
-data ObjectFile n = ObjectFile
-  { objectFileContents :: BS.ByteString
-  , objectFileFunsDefined :: [CFunName]
-  -- XXX: direct dependencies only
-  , objectFileDeps :: [ObjectFileName n] }
+-- Object files contain the object code for a single specialized function. All
+-- that remains to make it callable is to link it, load it into memory and
+-- obtain its pointer. But there's a mismatch between the way we handle names in
+-- the Dex compiler (well-scoped, alpha-renamable) vs the C/LLVM world (global,
+-- fixed). To reconcile the two approaches, we use C/LLVM names that are
+-- completely local to the object file itself, with no requirements for any sort
+-- of uniqueness beyond it. The function itself gets called
+-- `dexObjectFileMain_foo`, and the Dex functions it calls are called
+-- `dexCalledFunction_0_bar`, `dexCalledFunction_1_baz` (The `_foo` suffixes are
+-- just for readability and don't need to be unique.) The only way we use these
+-- names is to link the object code against known pointers to the functions it
+-- calls, and then to obtain its own pointer.
+
+type LocalCName = String -- has no meaning beyond a single LLVM module or object file
+data CNameMap (e::E) (n::S) = CNameMap
+  { mainFunName    :: LocalCName
+  , calledFunNames :: M.Map LocalCName (e n)
+  , dtorList       :: [LocalCName] }
```

Comment:
It feels weird to have the main func name and destructors be part of a "name map", since they don't map to anything. Perhaps we should move them to some extra fields of a function object code and keep this a map for just its dependencies